File: cmd/agent/ipsec.go
package main

import (
	"encoding/hex"
	"fmt"
	"log"
	"net"
	"os/exec"
	"strings"
)

type IPSecManager struct {
	spi      uint32
	key      []byte
	localIP  net.IP
	remoteIP net.IP
}

func NewIPSecManager(spi uint32, keyHex string, localIP, remoteIP net.IP) (*IPSecManager, error) {
	key, err := hex.DecodeString(keyHex)
	if err != nil {
		return nil, fmt.Errorf("invalid IPsec key: %w", err)
	}
	if len(key) != 16 && len(key) != 32 {
		return nil, fmt.Errorf("IPsec key must be 16 or 32 bytes, got %d", len(key))
	}
	return &IPSecManager{
		spi:      spi,
		key:      key,
		localIP:  localIP,
		remoteIP: remoteIP,
	}, nil
}

func (m *IPSecManager) AddSA() error {
_spi := fmt.Sprintf("0x%08x", m.spi)
_spiHex := hex.EncodeToString(m.key)

	// Use AES-GCM for encryption
	cmd := exec.Command("ip", "xfrm", "state", "add",
		"src", m.localIP.String(), "dst", m.remoteIP.String(),
		"proto", "esp",
		"spi", _spi,
		"enc", "aes-gcm", _spiHex,
		"mode", "tunnel",
		"replay-window", "128",
	)
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("add SA: %w: %s", err, string(out))
	}

	// Add policy
	cmd = exec.Command("ip", "xfrm", "policy", "add",
		"src", m.localIP.String()+"/32",
		"dst", m.remoteIP.String()+"/32",
		"dir", "out",
		"tmpl",
		"src", m.localIP.String(),
		"dst", m.remoteIP.String(),
		"proto", "esp",
		"spi", _spi,
		"mode", "tunnel",
	)
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("add policy: %w: %s", err, string(out))
	}

	log.Printf("[IPsec] added SA: %s <-> %s (SPI: 0x%08x)", m.localIP, m.remoteIP, m.spi)
	return nil
}

func (m *IPSecManager) DelSA() error {
	_spi := fmt.Sprintf("0x%08x", m.spi)

	cmd := exec.Command("ip", "xfrm", "state", "del",
		"src", m.localIP.String(), "dst", m.remoteIP.String(),
		"proto", "esp",
		"spi", _spi,
	)
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("del SA: %w: %s", err, string(out))
	}

	cmd = exec.Command("ip", "xfrm", "policy", "del",
		"src", m.localIP.String()+"/32",
		"dst", m.remoteIP.String()+"/32",
		"dir", "out",
	)
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("del policy: %w: %s", err, string(out))
	}

	log.Printf("[IPsec] removed SA: %s <-> %s", m.localIP, m.remoteIP)
	return nil
}

func (m *IPSecManager) RotateKey(newKeyHex string) error {
	newKey, err := hex.DecodeString(newKeyHex)
	if err != nil {
		return fmt.Errorf("invalid new key: %w", err)
	}

	// Delete old SA
	_ = m.DelSA()

	// Update key
	m.key = newKey
	m.spi++

	// Add new SA
	return m.AddSA()
}

func (m *IPSecManager) DumpStates() (string, error) {
	cmd := exec.Command("ip", "xfrm", "state", "list")
	out, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("dump states: %w", err)
	}
	return strings.TrimSpace(string(out)), nil
}

File: agent/src/ipsec.rs
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::time::Duration;

use anyhow::Result;
use tokio::process::Command;

use crate::AgentState;

pub struct IpsecManager {
    pub spi: u32,
    pub key: Vec<u8>,
}

impl IpsecManager {
    pub fn new(spi: u32, key: &str) -> Self {
        Self {
            spi,
            key: key.as_bytes().to_vec(),
        }
    }

    pub async fn add_state(&self, src: &str, dst: &str) -> Result<()> {
        let key_hex = hex::encode(&self.key);

        // Add inbound state
        let status = Command::new("ip")
            .args([
                "xfrm", "state", "add",
                "src", src,
                "dst", dst,
                "proto", "esp",
                "spi", &format!("{:08x}", self.spi),
                "reqid", &format!("{}", self.spi),
                "mode", "tunnel",
                "auth", "sha256", &key_hex,
                "enc", "aes", &key_hex,
            ])
            .status()
            .await?;
        if !status.success() {
            anyhow::bail!("ip xfrm state add (in) failed for {} -> {}", src, dst);
        }

        // Add outbound state
        let status = Command::new("ip")
            .args([
                "xfrm", "state", "add",
                "src", dst,
                "dst", src,
                "proto", "esp",
                "spi", &format!("{:08x}", self.spi),
                "reqid", &format!("{}", self.spi),
                "mode", "tunnel",
                "auth", "sha256", &key_hex,
                "enc", "aes", &key_hex,
            ])
            .status()
            .await?;
        if !status.success() {
            anyhow::bail!("ip xfrm state add (out) failed for {} -> {}", src, dst);
        }

        tracing::info!("IPsec state added: {} <-> {}", src, dst);
        Ok(())
    }

    pub async fn add_policy(&self, src_subnet: &str, dst_subnet: &str, tunnel_src: &str, tunnel_dst: &str) -> Result<()> {
        let status = Command::new("ip")
            .args([
                "xfrm", "policy", "add",
                "src", src_subnet,
                "dst", dst_subnet,
                "dir", "out",
                "tmpl", "src", tunnel_src,
                "dst", tunnel_dst,
                "proto", "esp",
                "mode", "tunnel",
                "reqid", &format!("{}", self.spi),
            ])
            .status()
            .await?;
        if !status.success() {
            anyhow::bail!("ip xfrm policy add (out) failed for {} -> {}", src_subnet, dst_subnet);
        }

        let status = Command::new("ip")
            .args([
                "xfrm", "policy", "add",
                "src", dst_subnet,
                "dst", src_subnet,
                "dir", "in",
                "tmpl", "src", tunnel_dst,
                "dst", tunnel_src,
                "proto", "esp",
                "mode", "tunnel",
                "reqid", &format!("{}", self.spi),
            ])
            .status()
            .await?;
        if !status.success() {
            anyhow::bail!("ip xfrm policy add (in) failed for {} -> {}", dst_subnet, src_subnet);
        }

        tracing::info!("IPsec policy added: {} <-> {}", src_subnet, dst_subnet);
        Ok(())
    }

    pub async fn delete_state(&self, src: &str, dst: &str) -> Result<()> {
        let _ = Command::new("ip")
            .args(["xfrm", "state", "del", "src", src, "dst", dst, "spi", &format!("{:08x}", self.spi), "proto", "esp"])
            .status().await;
        let _ = Command::new("ip")
            .args(["xfrm", "state", "del", "src", dst, "dst", src, "spi", &format!("{:08x}", self.spi), "proto", "esp"])
            .status().await;
        Ok(())
    }

    pub async fn delete_policy(&self, src_subnet: &str, dst_subnet: &str, dir: &str) -> Result<()> {
        let _ = Command::new("ip")
            .args(["xfrm", "policy", "del", "src", src_subnet, "dst", dst_subnet, "dir", dir])
            .status().await;
        Ok(())
    }

    pub async fn rotate_key(&mut self, new_key: &str) -> Result<()> {
        // Store old key for dual-SA rotation
        let old_spi = self.spi;
        let _old_key = self.key.clone();

        // Update SPI to avoid state conflicts
        self.spi = self.spi.wrapping_add(1);
        self.key = new_key.as_bytes().to_vec();

        tracing::info!("IPsec key rotation: old spi={:08x} new spi={:08x}", old_spi, self.spi);
        Ok(())
    }
}

pub async fn ipsec_loop(
    state: Arc<AgentState>,
    mut ipsec: Arc<std::sync::Mutex<IpsecManager>>,
    stop: Arc<AtomicBool>,
) {
    if !state.cfg.ipsec_enabled {
        return;
    }

    let mut interval = tokio::time::interval(Duration::from_secs(3600)); // Rotate every hour
    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);

    loop {
        if stop.load(Ordering::SeqCst) { return; }
        interval.tick().await;
        if stop.load(Ordering::SeqCst) { return; }

        {
            let mut ipsec = ipsec.lock().unwrap();
            // Generate a new key - for now use a deterministic one
            let new_key = format!("key-rotated-{}", std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap_or_default()
                .as_secs());
            if let Err(e) = ipsec.rotate_key(&new_key).await {
                tracing::warn!("IPsec key rotation failed: {}", e);
            }
        }

        // TODO: re-apply XFRM states with new SPI for all known peers
        tracing::info!("IPsec key rotation completed");
    }
}
